### PR TITLE
 [hardware, vllm, trainer] feat: add GRPO RL training support with TorchTitan actor and vLLM rollout on TPU                       

### PR DESCRIPTION
**Pull request by jialei777**
### What does this PR do?

This PR enables GRPO RL training support for Google Cloud TPU v6e instances on GKE using TorchTitan actor engine and vLLM rollout backend.            
                                                                                                                                                          
Key capabilities introduced:
                                                                             
- **Non-colocated TPU Multi-Slice Support**: Dedicated placement group isolation for TorchTitan trainer (`Slice 0`) and vLLM sampler/rollout (`Slice  1`).                                                                                                                                                    
- **Domain-Based Placement Group Resolution**: Added worker domain name-prefix filtering (`matching_pgs`) in `PlatformTPU.get_tpu_env_vars()` to  prevent inter-slice PJRT mesh mixing.                                                                                                                   
 - **TPU Checkpoint Engine Integration**: Implemented custom zero-copy weight synchronization via `TPUCheckpointEngine` and   `load_weights_from_ray_registry()`.                                                                                                                     
- **Vectorized TPU Logprobs**: Vectorized logprobs calculation in `torch_functional.py` for 30x faster reference logprob computation.     

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

Validated end-to-end on a GKE TPU v6e-8 multi-slice cluster (2 nodes x 4 chips for trainer, 2 nodes x 4 chips for rollout) running 5 full GRPO  training steps on Qwen3-0.6B with GSM8K dataset:

- Step 0 compilation completed cleanly.
- Steps 1..5 completed in ~8–10s/step.
- Metrics logged: `actor/loss=0.00165`, `actor/entropy=0.65234`, `actor/grad_norm=0.77039`.

### API and Usage Example

```bash
ray job submit --address "${RAY_ADDRESS}" \
  --working-dir . \
  --runtime-env-json '{
    "excludes": [".git", "logs", "*.log", "*.pt", "*.bin"],
    "env_vars": {
      "PYTHONPATH": ".",
      "PYTHONUNBUFFERED": "1",
      "VERL_PLATFORM": "tpu",
      "VLLM_USE_V1": "0",
      "RAY_memory_monitor_refresh_ms": "0",
      "RAY_memory_usage_threshold": "0.99",
      "RAY_EXPERIMENTAL_NOSET_TPU_VISIBLE_CHIPS": "1",
      "RAY_OVERRIDE_JOB_RUNTIME_ENV": "1",
      "WANDB_API_KEY": "'"${WANDB_API_KEY}"'"
    }
  }' \
  -- bash examples/tpu/grpo/run_qwen3_0_6b_torchtitan.sh
```

### Design & Code Changes

  1. verl/plugin/platform/platform_tpu.py: Added matching_pgs filtering in get_tpu_env_vars() to isolate Slice 0 (Trainer) and Slice 1 (Rollout) PJRT meshes.
  2. verl/checkpoint_engine/tpu_checkpoint_engine.py: Custom TPU weight transfer and registry loading without CUDA dependencies.
  3. verl/workers/rollout/vllm_rollout/tpu_utils.py: Applied TPU patches for vLLM EngineCore child process environment propagation and placement group  bindings.
  4. examples/tpu/grpo/run_grpo_qwen3_0.6b.sh: Added end-to-end TPU GRPO launch script and documentation at examples/tpu/grpo/README.md.
  

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/utils/distributed.py [hunk 1]
```diff
@@ -84,6 +90,10 @@ def initialize_global_process_group_ray(timeout_second=None, backend=None):
 
     import torch.distributed
 
+    # On TPU platforms, import torch_tpu to register the TPU communication backend with torch.distributed.
+    if get_platform().device_name == "tpu":
+        pass
```
Comment (andrewsykim) [inline verl/utils/distributed.py:95]: what's the reason for passing here instead of trying to initialize torch.distributed using `tpu_dist` backend?

File: examples/tpu/grpo/README.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  --runtime-env-json '{
+    "excludes": [".git", "logs", "*.log", "*.pt", "*.bin"],
+    "env_vars": {
+      "PYTHONPATH": ".",
+      "PYTHONUNBUFFERED": "1",
+      "VERL_PLATFORM": "tpu",
+      "VLLM_USE_V1": "0",
+      "RAY_memory_monitor_refresh_ms": "0",
+      "RAY_memory_usage_threshold": "0.99",
+      "RAY_EXPERIMENTAL_NOSET_TPU_VISIBLE_CHIPS": "1",
```
Comment (andrewsykim) [inline examples/tpu/grpo/README.md:44]: I'm assuming we need to set `RAY_EXPERIMENTAL_NOSET_TPU_VISIBLE_CHIPS` to disable default behavior for TPU_VISIBLE_CHIPS from Ray. Would be good to understand why this is required based on how veRL expects TPU_VISIBLE_CHIPS to be configured

File: examples/tpu/grpo/README.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+# Submit GRPO RL training job
+ray job submit --address "${RAY_ADDRESS}" \
+  --working-dir . \
+  --runtime-env-json '{
+    "excludes": [".git", "logs", "*.log", "*.pt", "*.bin"],
+    "env_vars": {
+      "PYTHONPATH": ".",
+      "PYTHONUNBUFFERED": "1",
+      "VERL_PLATFORM": "tpu",
+      "VLLM_USE_V1": "0",
```
Comment (andrewsykim) [inline examples/tpu/grpo/README.md:41]: What's the reason for disable v1 core engine? VLLM_USE_V1=1 has been the default for quite a while and more recent versions of vLLM have removed this entirely.

Comment (andrewsykim) [reply examples/tpu/grpo/README.md:41]: See https://github.com/vllm-project/vllm/issues/18571 for reference

Comment (andrewsykim) [reply examples/tpu/grpo/README.md:41]: If we tested this with a newer vLLM version (>0.15.0), I think this field is ignored and we can just remove it. Otherwise we must be testing with a very old version of vLLM on the old (deprecated) engine

File: verl/plugin/platform/platform_tpu_workarounds.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        return tuple(convert_tensors_to_scalars(v) for v in val)
+    elif hasattr(val, "aggregate"):
+        try:
+            return val.aggregate()
+        except Exception:
+            return val
+    return val
+
+
+def patch_ray_worker():
```
Comment (andrewsykim) [inline verl/plugin/platform/platform_tpu_workarounds.py:62]: I have a suspicion that we do not need this if we unset RAY_EXPERIMENTAL_NOSET_TPU_VISIBLE_CHIPS, but let me test and validate this. Either way, patching of Ray private method is something we want to avoid in general

File: verl/plugin/platform/platform_tpu_workarounds.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+def run_trainer_on_tpu(trainer_cls, config):
+    """Executes the SFT/PPO trainer inside a remote Ray TaskRunner actor on TPU worker nodes.
+
+    The Ray driver process runs on the head/CPU node where TPU devices are not present.
+    Wrapping trainer initialization inside a remote actor ensures that PJRT client initialization
+    and device setup run directly on TPU worker nodes.
+    """
+
+    @ray.remote(num_cpus=DEFAULT_TASK_RUNNER_CPUS, max_concurrency=DEFAULT_TASK_RUNNER_CONCURRENCY)
```
Comment (andrewsykim) [inline verl/plugin/platform/platform_tpu_workarounds.py:171]: I would think that this trainer remote function needs to have a resource definition for TPUs. if it's not set here, where is it being set? I'm assuming we're mapping 1 TPU device per trainer actor.
